httpbp/response.go

```diff
@@ -0,0 +1,218 @@
+package httpbp
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"html/template"
+	"net/http"
+
+	httpgk "github.com/go-kit/kit/transport/http"
+)
+
+const (
+	// ContentTypeHeader is the 'Content-Type' header key.
+	ContentTypeHeader = "Content-Type"
+
+	// JSONContentType is the Content-Type header for JSON responses.
+	JSONContentType = "application/json; charset=utf-8"
+
+	// HTMLContentType is the Content-Type header for HTML responses.
+	HTMLContentType = "text/html; charset=utf-8"
+)
+
+// HTTPError is a specialized error that is returned be the Err method specified
+// in the ErrorResponse interface.
+type HTTPError struct {
+	// Code is the status code to set on the HTTP response.  Defaults to 500 if
+	// it is not set.
+	Code int
+
+	// Message is an optional message that can be returned to the
+	// client.  Defaults to the native http.StatusText message for the
+	// StatusCode() of the HTTPError.
+	Message string
+
+	// Cause is an optional error that can be used to retain the error that
+	// led to us returning an HTTP error to the client.
+	Cause error
+}
+
+// Error returns the standard error string, this is not returned to the client.
+func (e HTTPError) Error() string {
+	return fmt.Sprintf(
+		"httpbp: http error with code %d, message %q and cause %v",
+		e.Code,
+		e.Message,
+		e.Cause,
+	)
+}
+
+// ResponseMessage returns the error message to send to the client.
+func (e HTTPError) ResponseMessage() string {
+	if e.Message != "" {
+		return e.Message
+	}
+	return http.StatusText(e.StatusCode())
+}
+
+// As implements helper interface for errors.As.
+//
+// If v is pointer to either HTTPError or *HTTPError,
+// *v will be set into this error.
+// Otherwise, As will try errors.As against the Cause of the HTTPError
+func (e HTTPError) As(v interface{}) bool {
+	if target, ok := v.(*HTTPError); ok {
+		*target = e
+		return true
+	} else if target, ok := v.(**HTTPError); ok {
+		*target = &e
+		return true
+	}
+	return errors.As(e.Cause, v)
```

Comment:
🔕 This part is kinda unnecessary. The whole purpose of `As` is to be used by `errors.As`. If you look at [its implementation](https://github.com/golang/go/blob/7d2473dc81c659fba3f3b83bc6e93ca5fe37a898/src/errors/wrap.go#L80-L87), it basically does these 3 things, in order:

1. try cast err into the type of `*v` directly (which is half of what we did above, but we try both pointer and non-pointer cases)
2. call `err`'s `As` implementation, if it's implemented (this function)
3. unwrap `err` and go back to 1

So if we just return false here, step 3 will still take care of it (because in our `Unwrap` implementation we just return `e.Cause`), with a slightly shallower call stack :)

Reply:
Note: if we don't implement `As`, we can achieve the same thing with trying to call `errors.As` against both `HTTPError` and `*HTTPError` (or only implement error interface in `*HTTPError` to eliminate the possibility of using `HTTPError`):

```
var he HTTPError
if errors.As(err, &he) {
  ...
}
var phe *HTTPError
if errors.As(err, &phe) {
  ...
}
```